app/uk/gov/hmrc/gform/gform/SummaryExprSubstituter.scala

```diff
@@ -123,42 +125,88 @@ object SummarySubstituter {
     }
 
   case class SummarySubstitutions(
-    exprMap: Map[Expr, ExpressionResult]
+    exprMap: Map[Expr, ExpressionResult],
+    data: VariadicFormData[SourceOrigin.OutOfDate]
   ) {
 
-    /** This method is used to replace the Sum expression with Adds
-      * It also substitutes the hidden FormCtx in the sumExpr with Constant(0)
-      */
     def replaceSumWithAdds(sumExpr: Expr): Expr = {
-      val sumExprFcIds = sumExpr.allFormComponentIds()
-      val indexedSumFcIdMap = exprMap.collect {
-        case (FormCtx(fcId), evalResult)
-            if sumExprFcIds
-              .map(_.baseComponentId)
-              .contains(fcId.baseComponentId) && fcId.modelComponentId.indexedComponentId.isIndexed =>
-          fcId -> evalResult
-      }.toMap
-
-      val indexedSumExprFcIds = indexedSumFcIdMap.map(_._1)
-      val indices = indexedSumExprFcIds
-        .map(fcId => fcId.modelComponentId.indexedComponentId)
-        .collect { case IndexedComponentId.Indexed(_, index) =>
-          index
-        }
-        .toList
-        .sorted
-        .distinct
+      val atlHelper = new ATLNestedAddHelper(exprMap, sumExpr)
+      val groupHelper = new GroupNestedAddHelper(data, sumExpr)
+      if (atlHelper.indexedFormComponentIds.nonEmpty) {
+        atlHelper.toNestedAdd()
+      } else if (groupHelper.indexedFormComponentIds.nonEmpty) {
+        groupHelper.toNestedAdd()
+      } else {
+        sumExpr
+      }
+    }
+  }
+
+  abstract class NestedAddHelper(sumExpr: Expr) {
+
+    val sumExprFcIds = sumExpr.allFormComponentIds()
 
+    def computeIndexedFormComponentIds(): List[FormComponentId]
+    def computeIndices(): List[Int]
+
+    lazy val indexedFormComponentIds: List[FormComponentId] = computeIndexedFormComponentIds()
```

Comment:
because I might never use it. with my if .. else ... else ... block above, I choose either ATL or Group helper to substitute. I don't what to do unnessasary computation on the helper constructor